src/main/java/com/prgrms/springbootbasic/customer/storage/JdbcCustomerRepository.java

```diff
@@ -0,0 +1,124 @@
+        DELETE("delete from customer where customer_id = UNHEX(REPLACE(?, '-', ''))"),
+        DELETE_BY_NAME("delete from customer where name = ?");
+
+        private final String query;
+
+        CustomerQuery(String query) {
+            this.query = query;
+        }
+    }
+
+    private enum CustomerColumn {
+        ID("customer_id"),
+        NAME("name"),
+        CREATED_AT("created_at");
+
+        private final String column;
+
+        CustomerColumn(String column) {
+            this.column = column;
+        }
+    }
+
+    private final JdbcTemplate jdbcTemplate;
+
+    public JdbcCustomerRepository(DataSource dataSource) {
+        jdbcTemplate = new JdbcTemplate(dataSource);
+    }
+
+    private static final RowMapper<Customer> ROW_MAPPER = (resultSet, rowNum) -> {
+        String name = resultSet.getString(CustomerColumn.NAME.column);
+        UUID customerId = toUUID(resultSet.getBytes(CustomerColumn.ID.column));
+        LocalDateTime createAt = resultSet.getTimestamp(CustomerColumn.CREATED_AT.column).toLocalDateTime();
+        return new Customer(customerId, createAt, name);
+    };
+
+    private static UUID toUUID(byte[] bytes) {
+        ByteBuffer wrap = ByteBuffer.wrap(bytes);
+        return new UUID(wrap.getLong(), wrap.getLong());
+    }
+
+    @Override
+    public void save(Customer customer) {
+        jdbcTemplate.update(CustomerQuery.INSERT.query,
+                customer.getId().toString(),
+                customer.getName(),
+                Timestamp.valueOf(customer.getCreatedAt()));
+    }
+
+    @Override
+    public void update(Customer customer) {
+        int update = jdbcTemplate.update(CustomerQuery.UPDATE.query,
+                customer.getName(),
+                customer.getId().toString());
+        if (update == 0) {
+            throw new DataModifyingException(
+                    "Nothing was updated. query: " + CustomerQuery.UPDATE.query + " params: " + customer.getName() + ", " + customer.getId());
+        }
+    }
+
+    @Override
+    public void delete(UUID id) {
+        int update = jdbcTemplate.update(CustomerQuery.DELETE.query,
+                id.toString());
+        if (update == 0) {
+            throw new DataModifyingException("Nothing was deleted. query: " + CustomerQuery.DELETE.query + " params: " + id);
+        }
+    }
+
+    @Override
+    public void delete(String name) {
+        int update = jdbcTemplate.update(CustomerQuery.DELETE_BY_NAME.query, name);
+        if (update == 0) {
+            throw new DataModifyingException(
+                    "Nothing was deleted. query: " + CustomerQuery.DELETE_BY_NAME.query + " params: " + name);
+        }
+    }
+
+    @Override
+    public Optional<Customer> findById(UUID id) {
+        return Optional.ofNullable(jdbcTemplate.queryForObject(
+                CustomerQuery.FIND_BY_ID.query,
+                ROW_MAPPER,
+                id.toString()));
```

Comment:
```suggestion
        return Optional.of(jdbcTemplate.queryForObject(
                CustomerQuery.FIND_BY_ID.query,
                ROW_MAPPER,
                id.toString()));
```
queryForObject null safety한걸로 알고있는데요. 확정적인 of를 쓰는게 더 좋을 거 같아요.

Reply:
IDE가 추천해주길래 그런갑다 했는데,,, null safety하군요! 
queryForObject 코드를 보니까 element가 하나일때만 정상값을 반환하고 그 이외에는 예외를 던지는 것 같아요. 이런 예외처리때문에 null safety하다고 하신거죠?

Reply:
네 제가 생각한부분과 맞습니다. 근데 왜 Nullable 어노테이션이 달려있는지 🧐